src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -0,0 +1,663 @@
+          if (umiValue.nonEmpty) samRecord.setAttribute(umiTag, molecularBarcodes.mkString(SampleBarcodeDelimiter))
+
+          if (multiUmiTags.nonEmpty) {
+            val numMolecularBarcodeSegments = getNumMolecularBarcodes(fastqRecords)
+            if (numMolecularBarcodeSegments == multiUmiTags.length) {
+              multiUmiTags.zip(molecularBarcodes).foreach { case (tag, barcode) => samRecord.setAttribute(tag, barcode) }
+            }
+            else if (multiUmiTags.length == 1) {
+              samRecord.setAttribute(multiUmiTags.head, molecularBarcodes.mkString(SampleBarcodeDelimiter))
+            }
+            else { // more than one tag, but not the same as the number of molecular barcodes
+              unreachable(s"Did not find only one SAM tag or the same # of SAM tags for the molecular barcodes (${multiUmiTags.length} != $numMolecularBarcodeSegments)")
+            }
+          }
+          Some(samRecord)
+        }
+      }
+      AssignedRecords(sampleOrdinal=sampleOrdinal, records=records)
+    }
+
+    def getBarcodeMetrics: Iterable[SampleBarcodeMetric] = {
+      SampleBarcodeMetric.finalizeMetrics(this.barcodeToMetrics, this.noMatchBarcode)
+      this.barcodeToMetrics.values
+    }
+
+    private def noMatchBarcodeMetric: SampleBarcodeMetric = this.barcodeToMetrics(noMatchBarcode)
+
+    /** Gets the # of mismatches for the given sample and observed sample barcode read bases. */
+    private def sampleOrdinalAndMismatches(sample: Sample, sampleBarcodeReadBases: String): (Int, Int) = {
+      val sampleBarcode = sample.sampleBarcode match {
+        case Some(barcode) => barcode
+        case None => throw new IllegalArgumentException(s"Sample barcode required for sample with id '${sample.sampleId}'")
+      }
+      val forwardMismatches = countMismatches(sampleBarcodeReadBases.getBytes, sampleBarcode.barcodeBytes)
+      val numMismatches = if (matchReverseComplement) {
+        if (!sampleBarcode.isInstanceOf[CustomSampleBarcode]) unreachable("Cannot match the reverse complement unless the barcode is of type 'CustomSampleBarcode'.")
+        val reverseMismatches = countMismatches(sampleBarcodeReadBases.getBytes, sampleBarcode.asInstanceOf[CustomSampleBarcode].reverseComplementedBarcodeBytes)
+        require(reverseMismatches != 0 || forwardMismatches != 0) // otherwise they are reverse complements of each other
+        if (forwardMismatches < reverseMismatches) forwardMismatches
+        else reverseMismatches
+      }
+      else {
+        forwardMismatches
+      }
+      (sample.sampleOrdinal, numMismatches)
+    }
+
+    /** Searches for a matching sample given the observed barcode.  Returns -1 if no match is found, given the
+      * sample sheet and matching parameters */
+    private def getSampleOrdinalFromSampleBarcode(sampleBarcodeReadBases: String, sampleSheet: SampleSheet): Option[Int] = {
+      val numNoCalls: Int = sampleBarcodeReadBases.getBytes.count(base => SequenceUtil.isNoCall(base))
```

Comment:
This is, unfortunately, going to be slow due to `byte -> Byte -> byte` conversions.